src/Features/Core/Portable/RemoveUnusedVariable/AbstractRemoveUnusedVariableCodeFixProvider.cs

```diff
@@ -22,85 +24,157 @@ internal abstract class AbstractRemoveUnusedVariableCodeFixProvider<TLocalDeclar
-                        var removeOptions = SyntaxGenerator.DefaultRemoveOptions;
+                ISymbol symbol = documentEditor.SemanticModel.GetDeclaredSymbol(node);
+                var referencedSymbols = await SymbolFinder.FindReferencesAsync(symbol, document.Project.Solution, DocumentsToBeSearched, cancellationToken).ConfigureAwait(false);
 
-                        if (localDeclaration.GetLeadingTrivia().Contains(t => t.IsDirective))
-                        {
-                            removeOptions |= SyntaxRemoveOptions.KeepLeadingTrivia;
-                        }
-                        else
+                foreach (var referencedSymbol in referencedSymbols)
+                {
+                    if (referencedSymbol?.Locations != null)
+                    {
+                        foreach (var location in referencedSymbol.Locations)
                         {
-                            var statementParent = localDeclaration.Parent;
-                            if (syntaxFacts.IsExecutableBlock(statementParent))
+                            var referencedSymbolNode = root.FindNode(location.Location.SourceSpan);
+                            if (referencedSymbolNode != null)
                             {
-                                var siblings = syntaxFacts.GetExecutableBlockStatements(statementParent);
-                                var localDeclarationIndex = siblings.IndexOf(localDeclaration);
-                                if (localDeclarationIndex != 0)
+                                var nodeToRemoveOrReplace = GetNodeToRemoveOrReplace(referencedSymbolNode);
+                                if (nodeToRemoveOrReplace != null)
                                 {
-                                    // if we're removing hte first statement in a block, then we
-                                    // want to have the elastic marker on it so that the next statement
-                                    // properly formats with the space left behind.  But if it's
-                                    // not the first statement then just keep the trivia as is
-                                    // so that the statement before and after it stay appropriately
-                                    // spaced apart.
-                                    removeOptions &= ~SyntaxRemoveOptions.AddElasticMarker;
+                                    nodesToRemove.Add(nodeToRemoveOrReplace);
                                 }
                             }
                         }
-
-                        editor.RemoveNode(localDeclaration, removeOptions);
                     }
-                    else if (variableDeclarators.Count() > 1)
+                }
+            }
+
+            MergeNodesToRemove(nodesToRemove);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            foreach (var node in nodesToRemove.Where(n => n != null))
```

Comment:
we have a .WhereNotNull extenion for this.

Reply:
Not certain how any of these nodes could be null.

Reply:
FindNode could return null due some mismatch.

---
In reply to: [160015979](https://github.com/dotnet/roslyn/pull/24077#discussion_r160015979) [](ancestors = 160015979)

Reply:
Could you be more specific?  In what circumstances could you get null?

Reply:
I think that 
var node = root.FindNode(diagnostic.Location.SourceSpan);
could return null. For example, diagnostic.Location could be incorrect, or root and diagnostic could be inconsistent (refer to different snapshot). 

---
In reply to: [160250916](https://github.com/dotnet/roslyn/pull/24077#discussion_r160250916) [](ancestors = 160250916)

Reply:
> ```root.FindNode(diagnostic.Location.SourceSpan);``` could return null

This should not be possible.  Either SourceSpan is not in the bounds of the root (in which case an exception is thrown), or you should get a node (at the very least the CompilationUnit node).

> For example, diagnostic.Location could be incorrect, or root and diagnostic could be inconsistent (refer to different snapshot).

This shoudl not be possible either.  Even if diagnostic.Location was incrrrect, you'd get a node. And you should not be able to get inconsistency (otherwise something in the analyzer subsystem is broken).

Reply:
OK. Agreed. Thank you! Removing

---
In reply to: [160271426](https://github.com/dotnet/roslyn/pull/24077#discussion_r160271426) [](ancestors = 160271426)